### DISCUSSION TITLE
Modelspace area visible in the viewport

### DISCUSSION THREAD

**Discussion by gswifort** [CEXT-Dan/PyRx#101] [General]
The code below sometimes returns an incorrect `.viewCenter()` point. As I understand it, these should be the same values ​​as in the View section in the properties palette.
I am attaching a file where this problem occurs.

```py
from pyrx_imp import Db, Ed, Ge


def PyRxCmd_doit():
    status, ent_id, *_ = Ed.Editor.entSel("Select viewport: ")
    if not status == Ed.PromptStatus.eOk:
        return
    vp = Db.Viewport(ent_id)
    coords = (*vp.viewCenter().toTuple(), 0.0)
    print(coords)
    point = Db.Point(Ge.Point3d(*coords))
    Db.curDb().addToModelspace(point)
```

![image](https://github.com/user-attachments/assets/26e7b1b2-9731-405e-a7c9-966093e4ad47)
[viewport_bug.zip](https://github.com/user-attachments/files/16411891/viewport_bug.zip)

**Comment by CEXT-Dan** [comment]
viewCenter is in DCS, so you will have to translate the coordinates
https://help.autodesk.com/view/OARX/2024/ENU/?guid=GUID-01A45BA0-CC4F-4DCA-840E-DCA8802A060A
you will probably want from = 2, and to = 0(WCS) modelspace

```Python
from pyrx_imp import Ap, Db, Ed, Ge, Gi, Gs, Rx
import traceback

#function test
def PyRxCmd_doit():
    try:
        status, ent_id, *_ = Ed.Editor.entSel("Select viewport: ")
        if not status == Ed.PromptStatus.eOk:
            return
        vp = Db.Viewport(ent_id)
        coords = (*vp.viewCenter().toTuple(), 0.0)
        print(coords)
        for ele in Db.Core.entGetX(ent_id):
            if ele[0] == 1040:
                if ele[1] in coords and ele[1] > 0.0:
                    print(ele)
    except Exception as err:
        print(err)
        
        
def PyRxCmd_doit2():
    try:
        status, ent_id, *_ = Ed.Editor.entSel("Select viewport: ")
        if not status == Ed.PromptStatus.eOk:
            return
        vp = Db.Viewport(ent_id)
        cenDCS = Ge.Point3d(vp.viewCenter().toTuple()) #adds the zero for you
        
        rbfrom = [(Rx.LispType.kInt16, 2)] #DCS
        rbto = [(Rx.LispType.kInt16, 3)] #PSDCS
        
        cenPDCS = Ed.Core.trans(cenDCS,rbfrom,rbto,False)
        print(cenPDCS)
    except Exception as err:
        print(err)
```
![vcen](https://github.com/user-attachments/assets/e9aed829-6a08-404e-88b8-caaaa777a004)

**Comment by CEXT-Dan** [comment]
As a side note, Db.Core has ucs2Wcs, wcs2Ecs, wcs2Ucs, and ecs2Wcs. You must use acadtrans for DCS. It’s a handy function, for example, if you want the mouse position in WCS

**Comment by gswifort** [comment]
Your solution didn't work, but maybe because I didn't have this viewport active. The key seems to be `Viewport.viewTarget()`

```py
import traceback

from pyrx_imp import Db, Ed, Ge


def PyRxCmd_doit():
    try:
        status, ent_id, *_ = Ed.Editor.entSel("Select viewport: ")
        if not status == Ed.PromptStatus.eOk:
            return
        vp = Db.Viewport(ent_id)
        cenDCS = Ge.Point3d(vp.viewCenter().toTuple())
        cenWCS = cenDCS + Ge.Vector3d(*vp.viewTarget().toTuple()) # !!!
        print(cenWCS)
    except Exception:
        traceback.print_exc()
```

**Comment by CEXT-Dan** [comment]
fixing py_list_to_vector3d, missed that

**Comment by gswifort** [comment]
Another thing to think about is the viewport with a rotated Z axis.

_You probably want to convert this into a discussion :)_

**Reply by CEXT-Dan** [reply]
Gile has some excellent tools in .NET  
https://github.com/gileCAD/GeometryExtensions/blob/master/GeometryExtensionsR25/ViewportExtension.cs

```Python
import traceback
from pyrx_imp import Rx, Ge, Gs, Db, Ap, Ed

def DCS2WCS(vp: Db.Viewport):
    return (
        Ge.Matrix3d.rotation(-vp.twistAngle(), vp.viewDirection(), vp.viewTarget())
        * Ge.Matrix3d.translation(vp.viewTarget() - Ge.Point3d.kOrigin)
        * Ge.Matrix3d.planeToWorld(vp.viewDirection())
    )
    
def PyRxCmd_doit():
    try:
        status, ent_id, *_ = Ed.Editor.entSel("Select viewport: ")
        if not status == Ed.PromptStatus.eOk:
            return
        vp = Db.Viewport(ent_id)
        print(DCS2WCS(vp))
    except Exception as err:
        traceback.print_exception(err)

```

**Reply by gswifort** [reply]
Thanks! although I don't know how to use it yet :)
To clarify - my goal is to get what area of ​​the model is visible in the viewport. I assume that such information should be retrieved for a specific `Z` (if the coordinate system is rotated).

**Comment by gswifort** [comment]
https://github.com/CEXT-Dan/PyRx/blob/main/PySamples/PyDb/viewport.py
